osquery/tables/networking/windows/routes.cpp

```diff
@@ -0,0 +1,168 @@
+*  of patent rights can be found in the PATENTS file in the same directory.
+*
+*/
+
+#define WIN32_LEAN_AND_MEAN
+
+#include <memory>
+#include <string>
+#include <windows.h>
+#include <winsock2.h>
+
+#include <wS2tcpip.h>
+#include <ws2ipdef.h>
+
+#include <iphlpapi.h>
+#include <mstcpip.h>
+
+#include <boost/algorithm/string/join.hpp>
+#include <osquery/tables.h>
+
+#include "osquery/core/conversions.h"
+#include "osquery/core/windows/wmi.h"
+
+namespace osquery {
+namespace tables {
+
+std::map<unsigned long, std::shared_ptr<IP_ADAPTER_INFO>>
+getAdapterAddressMapping() {
+  std::map<unsigned long, std::shared_ptr<IP_ADAPTER_INFO>> returnMapping;
+  auto dwBufLen = 0UL;
+  auto dwStatus = GetAdaptersInfo(NULL, &dwBufLen);
+
+  if (dwStatus == ERROR_BUFFER_OVERFLOW) {
+    std::vector<BYTE> buffer(dwBufLen);
+    auto pAdapterInfo = reinterpret_cast<PIP_ADAPTER_INFO>(buffer.data());
+    dwStatus = GetAdaptersInfo(pAdapterInfo, &dwBufLen);
+
+    if (dwStatus != NO_ERROR) {
+      return returnMapping;
+    }
+
+    while (pAdapterInfo) {
+      auto sharedPtr = std::make_shared<IP_ADAPTER_INFO>(*pAdapterInfo);
+      returnMapping.insert(std::make_pair(pAdapterInfo->Index, sharedPtr));
+      pAdapterInfo = pAdapterInfo->Next;
+    }
+  }
+
+  return returnMapping;
+}
+
+PMIB_IPINTERFACE_TABLE getInterfaces(int type = AF_UNSPEC) {
+  auto interfaceTable = std::make_unique<PMIB_IPINTERFACE_TABLE>();
+
+  auto dwRetVal = GetIpInterfaceTable(type, interfaceTable.get());
+  if (dwRetVal != NO_ERROR) {
+    return nullptr;
+  }
+
+  return *interfaceTable.get();
+}
+
+std::map<unsigned long, MIB_IPINTERFACE_ROW> getInterfaceRowMapping(
+    int type = AF_UNSPEC) {
+  std::map<unsigned long, MIB_IPINTERFACE_ROW> returnMapping;
+  PMIB_IPINTERFACE_TABLE interfaces;
+
+  if ((interfaces = getInterfaces(type)) != nullptr) {
+    for (unsigned long i = 0; i < interfaces->NumEntries; ++i) {
+      MIB_IPINTERFACE_ROW currentRow = interfaces->Table[i];
+      returnMapping.insert(
+          std::make_pair(currentRow.InterfaceIndex, currentRow));
+    }
+  }
+
+  return returnMapping;
+}
+
+QueryData genRoutes(QueryContext& context) {
+  QueryData results;
+  PMIB_IPFORWARD_TABLE2* ipTable = nullptr;
+
+  ipTable = static_cast<PMIB_IPFORWARD_TABLE2*>(
+      malloc(sizeof(PMIB_IPFORWARD_TABLE2)));
+  auto result = GetIpForwardTable2(AF_UNSPEC, ipTable);
+
+  if (result != NO_ERROR) {
+    FreeMibTable(ipTable);
+
+    return results;
+  }
+
+  auto numEntries = ipTable[0]->NumEntries;
+  auto interfaces = getInterfaceRowMapping();
+  auto adapters = getAdapterAddressMapping();
+
+  for (unsigned long i = 0; i < numEntries; ++i) {
+    Row r;
+    std::string interfaceIpAddress;
+    PVOID ipAddress = nullptr;
+    PVOID gateway = nullptr;
+    auto currentRow = ipTable[0]->Table[i];
+    auto addrFamily = currentRow.DestinationPrefix.Prefix.si_family;
+    auto actualInterface = interfaces.at(currentRow.InterfaceIndex);
```

Comment:
If `currentRow.InterfaceIndex` does not exist within the `std::map` this will throw an exception.